fe/fe-core/src/main/java/org/apache/doris/qe/cache/CacheCoordinator.java

```diff
@@ -0,0 +1,162 @@
+import org.apache.doris.catalog.Catalog;
+import org.apache.doris.qe.SimpleScheduler;
+import org.apache.doris.system.Backend;
+import org.apache.doris.proto.PUniqueId;
+
+import com.google.common.collect.ImmutableMap;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import java.util.HashMap;
+import java.util.Hashtable;
+import java.util.List;
+import java.util.SortedMap;
+import java.util.TreeMap;
+import java.util.Iterator;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+/**
+ * Use consistent hashing to find the BE corresponding to the key to avoid the change of BE leading to failure to hit the Cache
+ */
+public class CacheCoordinator {
+    private static final Logger LOG = LogManager.getLogger(CacheCoordinator.class);
+    private static final int VIRTUAL_NODES = 10;
+    private static final int REFRESH_NODE_TIME = 300000;
+    public boolean DebugModel = false;
+    private Hashtable<Long, Backend> realNodes = new Hashtable<>();
+    private SortedMap<Long, Backend> virtualNodes = new TreeMap<>();
+    private static Lock belock = new ReentrantLock();
+
+    private long lastRefreshTime;
+    private static CacheCoordinator cachePartition;
+
+    public static CacheCoordinator getInstance() {
+        if (cachePartition == null) {
+            cachePartition = new CacheCoordinator();
+        }
+        return cachePartition;
+    }
+
+    protected CacheCoordinator() {
+    }
+
+    /**
+     * Using the consistent hash and the hi part of sqlkey to get the backend node
+     *
+     * @param sqlKey 128 bit's sql md5
+     * @return Backend
+     */
+    public Backend findBackend(PUniqueId sqlKey) {
+        checkBackend();
+        Backend virtualNode = null;
+        try {
+            belock.lock();
+            SortedMap<Long, Backend> headMap = virtualNodes.headMap(sqlKey.hi);
+            SortedMap<Long, Backend> tailMap = virtualNodes.tailMap(sqlKey.hi);
+            int retryTimes = 0;
+            while (true) {
+                if (tailMap == null || tailMap.size() == 0) {
+                    tailMap = headMap;
+                    retryTimes += 1;
+                }
+                Long key = tailMap.firstKey();
+                virtualNode = tailMap.get(key);
+                if (SimpleScheduler.isAlive(virtualNode)) {
+                    break;
+                } else {
+                    LOG.debug("backend {} not alive, key {}, retry {}", virtualNode.getId(), key, retryTimes);
+                    virtualNode = null;
+                }
+                tailMap = tailMap.tailMap(key + 1);
+                retryTimes++;
+                if (retryTimes >= 5) {
+                    LOG.warn("find backend, reach max retry times {}", retryTimes);
+                    break;
+                }
+            }
+        } finally {
+            belock.unlock();
+        }
+        return virtualNode;
+    }
+
+    public void checkBackend() {
```

Comment:
What's the meaning of `checkBackend`?  which `clearBackend` firstly, then `addBackend`. `rebuildBackend` or `resetBackend` seems better?